component/prometheus/source/api/api.go

```diff
@@ -0,0 +1,141 @@
+package api
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"reflect"
+	"sync"
+
+	"github.com/go-kit/log/level"
+	"github.com/gorilla/mux"
+	"github.com/grafana/agent/component"
+	fnet "github.com/grafana/agent/component/common/net"
+	agentprom "github.com/grafana/agent/component/prometheus"
+	"github.com/grafana/agent/pkg/util"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/storage/remote"
+)
+
+func init() {
+	component.Register(component.Registration{
+		Name: "prometheus.source.api",
+		Args: Arguments{},
+		Build: func(opts component.Options, args component.Arguments) (component.Component, error) {
+			return New(opts, args.(Arguments))
+		},
+	})
+}
+
+type Arguments struct {
+	Server    *fnet.ServerConfig   `river:",squash"`
+	ForwardTo []storage.Appendable `river:"forward_to,attr"`
+}
+
+type Component struct {
+	opts               component.Options
+	handler            http.Handler
+	fanout             *agentprom.Fanout
+	uncheckedCollector *util.UncheckedCollector
+
+	updateMut sync.RWMutex
+	args      Arguments
+	server    *fnet.TargetServer
+}
+
+func New(opts component.Options, args Arguments) (component.Component, error) {
+	fanout := agentprom.NewFanout(args.ForwardTo, opts.ID, opts.Registerer)
+
+	uncheckedCollector := util.NewUncheckedCollector(nil)
+	opts.Registerer.MustRegister(uncheckedCollector)
+
+	c := &Component{
+		opts:               opts,
+		handler:            remote.NewWriteHandler(opts.Logger, fanout),
+		fanout:             fanout,
+		uncheckedCollector: uncheckedCollector,
+	}
+
+	if err := c.Update(args); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// Run satisfies the Component interface.
+func (c *Component) Run(ctx context.Context) error {
+	defer func() {
+		c.updateMut.Lock()
+		defer c.updateMut.Unlock()
+		c.shutdownServer()
+	}()
+
+	for range ctx.Done() {
+		level.Info(c.opts.Logger).Log("msg", "terminating due to context done")
+		return nil
+	}
```

Comment:
Linter told me to use this syntax. Personally I don't mind a more verbose, but more intuitive version of
```
for {
  select {
    ...
  }
}
```
But seems that when there is only one `case` in the `select` statement, the `range` syntax is recommended.

Reply:
Why do we need the `for {}` here? 

Since there's nothing else on the processing loop (like we have for logging components), we could just block on `<-ctx.Done()`, right?